### ISSUE TITLE
Allow dry_run for snapshot_download

### ISSUE THREAD

**Issue by cccntu** [huggingface/huggingface_hub#1023]
The current download progress bar doesn't show the file names being downloaded, and doesn't show how many files will be downloaded
With allow_patterns , ignore_patterns, having a dry run option would be very useful.
One can also use it to add logging.

**Comment by Wauplin** [comment]
Hi @cccntu, thanks for the feature request :+1: 

Do you want to give it a try to implement it ? :)
Implementation of the dry-run option would most probably happen [somewhere here](https://github.com/huggingface/huggingface_hub/blob/c03bd2917c35b44d6a38a63dac4b0f72dad3ab99/src/huggingface_hub/_snapshot_download.py#L193). You might need to look into implement of [`hf_hub_download`](https://github.com/huggingface/huggingface_hub/blob/main/src/huggingface_hub/file_download.py#L881) as well to know which file is already cached from a previous revision and which file is new. `hf_hub_download` can only have the dry-run option.

If you need any help, I'd be happy to answer any question.

**Comment by druvdub** [comment]
Hey is it okay if I pick this issue? :)

**Comment by Wauplin** [comment]
Hi @druvdub, thanks for proposing you on this one! This issue is not yet assigned so yes, feel free to take it! :pray:  Implementation-wise, I think you can start with a PR that adds a `dry_run` option to `hf_hub_download` first. Adding support for `snapshot_download` and `huggingface-cli download` can be done in a follow-up PR. 

The biggest question is: what is the expected output of a download in dry-run mode? My expectation is that it would not download anything but it is still _ok-ish_ if it updates the internal refs (see these [docs](https://huggingface.co/docs/huggingface_hub/guides/manage-cache#refs)). As a return value, `None` should be fine and use logs to print what should have been done if it wasn't a dry-run. I see different cases:
- file already exists and at a correct place => nothing to do
- file already exists but not in correct place => requires only a symlink
- file doesn't exists in cache => requires a download
- if `resume_download=True` => how much left to download?

What do you think? Would you like to start by having a look at the code and then discuss what you think is a good direction to take? Happy to discuss it further if you're interested.

**Comment by druvdub** [comment]
@Wauplin Yes, I would most likely explore the codebase a bit and then we can discuss how to proceed forward on this issue

**Comment by druvdub** [comment]
@Wauplin right, did a bit of reading and was wondering if the output should be something parse able for the dry-run like in a JSON format along with the logs

**Comment by druvdub** [comment]
Sounds good. I'll make a draft PR with the changes.

**Comment by druvdub** [comment]
Hey @Wauplin, going through the code I found a HfFileMetadata class [here](https://github.com/huggingface/huggingface_hub/blob/0ab93911d902af16378a8f8d911320c9b03c1e65/src/huggingface_hub/file_download.py#L160), Would that be something we'd like to return or something other than that. I feel it might kind of be a duplicated class and we could reduce it to something generic or potentially extend this class

**Comment by Wauplin** [comment]
Hi @druvdub, thanks for looking into it! Actually I think it's best to start from a separate dataclass altogether, even if that means you end up with duplicated fields. We tried to play with inheritance in the past for those situations but it always went with drawbacks. Since we need to document all the attributes in the docstring anyway, we would not gain much. So let's have a dry-run -specific class :)

**Comment by vaibhav-research** [comment]
@druvdub @Wauplin is this still pending ?

**Comment by druvdub** [comment]
> @druvdub @Wauplin is this still pending ?

Hey, apologies. This completely slipped out of my mind. If this is still relevant and has not been picked up at all since then. I can pick it up and complete it this week.

**Comment by Wauplin** [comment]
Sorry @druvdub, your notifications slipped through my radar. Feature has been implemented in https://github.com/huggingface/huggingface_hub/pull/3407 and released in v1.0.
